src/WasmModule.ts

```diff
@@ -0,0 +1,206 @@
+import { basename, extname } from './utils/path';
+
+import ExternalModule from './ExternalModule';
+import Module from './Module';
+
+import ModuleScope from './ast/scopes/ModuleScope';
+import Variable from './ast/variables/Variable';
+import NamespaceVariable from './ast/variables/NamespaceVariable';
+
+const decoderOpts = {
+	ignoreCodeSection: true,
+	ignoreDataSection: true
+};
+
+type BuildLoadArgs = {
+	NAME: string;
+	URL: string;
+	IMPORT_OBJECT?: string;
+};
+
+const buildLoader = ({ NAME, URL, IMPORT_OBJECT }: BuildLoadArgs) => `
+	// function then$${NAME}(resolve) {
+	function then(resolve) {
+		if (typeof WebAssembly.instantiateStreaming !== 'function') {
+		  throw new Error('WebAssembly.instantiateStreaming is not supported');
+		}
+
+		if (typeof window.fetch !== 'function') {
+		  throw new Error('window.fetch is not supported');
+		}
+
+		const req = window.fetch('${URL}');
+
+		WebAssembly
+			.instantiateStreaming(req, ${IMPORT_OBJECT || '{}'})
+			.then(res => res.instance.exports)
+			.then(resolve)
+			.catch(resolve);
+	}
+`;
+
+export interface ExportDescription {
+	localName: string;
+}
+
+export interface ImportDescription {
+	source: string;
+	name: string;
+	module: Module | ExternalModule | null;
+}
+
+export default class WasmModule {
+	type: 'WasmModule';
+
+	id: string;
+	graph: Graph;
+	code: Buffer;
+
+	scope: ModuleScope;
+
+	ast: ast.Program;
+
+	sources: string[];
+	resolvedIds: IdMap;
+
+	dynamicImportResolutions: {
+		alias: string;
+		resolution: Module | ExternalModule | string | void;
+	}[];
+	dependencies: (Module | ExternalModule | WasmModule)[];
+
+	imports: { [name: string]: ImportDescription };
+	exports: { [name: string]: ExportDescription };
+
+	// this is unused on Module,
+	// only used for namespace and then ExternalExport.declarations
+	declarations: {
+		'*'?: NamespaceVariable;
+		[name: string]: Variable | undefined;
+	};
+
+	isExternal: false;
+
+	constructor(graph: Graph, id: string) {
+		this.id = id;
+		this.graph = graph;
+		this.code = new Buffer('');
+
+		this.ast = null;
+
+		// imports and exports, indexed by local name
+		this.imports = Object.create(null);
+		this.exports = Object.create(null);
+		this.resolvedIds = Object.create(null);
+		this.declarations = Object.create(null);
+
+		// all dependencies
+		this.dynamicImportResolutions = [];
+		this.sources = [];
+		this.dependencies = [];
+
+		this.scope = new ModuleScope(<any>this);
+
+		// expose Thenable which is the entry point of our loader
+		// this.exports['then$' + this.basename()] = {
+		this.exports.then = {
+			localName: 'then'
+		};
+
+		// FIXME(sven): a different then allows multiple wasm modules to be load
+		// in the same chunk (avoids collision). I need to figure out how to
+		// have an object like {then: then$foo} as the export;
+	}
+
+	render() {
+		const NAME = this.basename();
+		const URL = `/dist/${NAME}.wasm`;
+
+		const content = buildLoader({ URL, NAME });
+
+		return { trim() {}, content };
+	}
```

Comment:
If we have WasmChunk calling render, then it will be ok for the render method to return a binary (plus binary source map in due course).